<a href="https://colab.research.google.com/github/JBadawood/Mathematical_Optimization/blob/main/Linear_Programming/What_Is_LP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **What is Linear Programming (LP)?**

[What is Linear Programming (LP)? (in 2 minutes)](https://youtu.be/C0TTxV0n9OA)

CVXPY is an open source Python-embedded modeling language for **convex optimization** problems.\
https://www.cvxpy.org/

[GitHub](https://github.com/cvxpy/cvxpy)

---
**تحسين محدب**\
عملية إيجاد الحد الأدنى من **دالة محدبة** باستخدام أساليب رياضية.

**Convex Optimization**\
A process of finding the minimum of a **convex function** using mathematical techniques.

--

**دالة محدبة**\
دالة تكون المنطقة فوق رسمها البياني عبارة عن **مجموعة محدبة**.

**Convex Function**\
A function in which the area above its graph is a **convex set**.

--

**مجموعة محدبة**\
مجموعة فرعية من الفضاء الإقليدي حيث يبقى الخط المرسوم بين أي نقطتين في المجموعة الفرعية بالكامل داخل المجموعة الفرعية.

**Convex Set**\
A subset of Euclidean space in which a drawn line between any two points in the subset stays totally within the subset.

>المصدر: [معجـــم البيانـــات والـــذكاء االصطناعـــي](https://sdaia.gov.sa/files/Dictionary.pdf) من الهيئة السعودية للبيانات والذكاء الاصطناعي بالتعاون مع مجمع الملك سلمان العالمي للغة العربية (الطبعة الأولى 2022)
--- 

CVXPY **is not** a solver. It relies upon the open source solvers ECOS, SCS, and OSQP. Additional solvers are available, but must be installed separately.

\

**Choosing a solver**\
https://www.cvxpy.org/tutorial/advanced/index.html#choosing-a-solver
\
CVXPY is distributed with the open source solvers ECOS, OSQP, and SCS. Many other solvers can be called by CVXPY if installed separately:
- [Google's Linear Optimization Package (GLOP)](https://github.com/google/or-tools/tree/stable/ortools/glop) | [Goolgle OR-Tools](https://developers.google.com/optimization)
- CBC
- GLPK
- GLPK_MI
- OSQP
- PDLP
- CPLEX
- NAG
- ECOS
- GUROBI
- MOSEK
- CVXOPT
- SCS
- SCIP
- XPRESS
- SCIPY

## **1- Example of Linear Programming**

لو أننا في شركة تنتج نوعين من السيارات الكهربائية:
- السيارة الأولى من نوع S تتطلب يومين عشان عشان نسويها وأربع بطاريات وهامش الربح فيها ألف ريال
- أما السيارة من نوع y راح نصنعها في 15 يوم وتتطلب 8 بطاريات وبهامش ربح خمسة آلاف ريال

لو أعطينا مدة سنة كاملة ومخزون من 500 بطارية كيف ممكن نزود من أرباح الشركة؟

Linear Programming

Decision Variables:\
$s$: عدد السيارات من موديل s\
$y$: عدد السيارات من موديل y

\

Objective Function:\
$Max:\ \ z=1000*s + 5000*y$

subject to:\
$4*s + 8*y \le 500$

$2*s + 15*y \le 365$

> نلاحظ أن كلا دالة الهدف والقيود عبارة عن دوال خطية لذلك راح نسميها (برمجة خطية)


The set of all decisions that satisfy the constraints is called the **Feasible Set**.

راح نعرف البرمجة الخطية على أنها تتعامل مع faces of feasible set وكأنها falt وهذا ما يعرف بمصطلح **Polyhedron** وهي عبارة عن الـ set with flat faces

https://youtu.be/C0TTxV0n9OA?t=40

نقدر نوجد الـ optimal solution بطريقتين:
- Simplex Method
- Interior Point Method

## **2- Solving Linear Program in Python**

### 2-1 Install & Import CVXPY

In [1]:
# Install the package cvxpy
!pip install cvxpy

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [2]:
# Import cvxpy
import cvxpy as cp

### 2-2 Describe the Problem

#### 2-2-1 Decision Varibles

Decision Variables:\
$s$: عدد السيارات من موديل s\
$y$: عدد السيارات من موديل y

In [5]:
# Decision Varibles
s = cp.Variable()
y = cp.Variable()

#### 2-2-2 Objective Function

Objective Function:\
$Max:\ \ z=1000*s + 5000*y$

In [6]:
# Objective Function
objective = cp.Maximize(1000*s + 5000*y)

#### 2-2-3 Constraints

subject to:\
$4*s + 8*y \le 500$

$2*s + 15*y \le 365$

In [8]:
# Constraints
constraints = [
    4*s + 8*y <=500,
    2*s + 15*y <= 365
]

### 2-3 Call the Solve Method & Optimal Solution

#### 2-3-1 Solve Method

In [9]:
# Solve method
prob = cp.Problem(objective, constraints)

In [10]:
prob

Problem(Maximize(Expression(AFFINE, UNKNOWN, ())), [Inequality(Expression(AFFINE, UNKNOWN, ())), Inequality(Expression(AFFINE, UNKNOWN, ()))])

#### 2-3-2 Optimal Solution

In [11]:
# Get the optimal solution
optimal_value = prob.solve()

In [12]:
optimal_value

156363.6363638718

In [13]:
s.value

array(104.09090909)

In [14]:
y.value

array(10.45454545)

> بكذا نعرف أن الأربح ممكن نوصلها أفضل شي إلى أن تصل قرابة 156 ألف ريال لمن يكون عدد السيارات من موديل s 104 سيارة ومن موديل y 10 سيارات

نلاحظ أن الأرقام بالكسور لأننا استخدمنا Linear programming بدل الـ integer programming

In [18]:
print(f"Maximum Profit = {optimal_value},\nwhen:\nNo. of s model is {s.value}\nand\nNo. of y model is {y.value}")

Maximum Profit = 156363.6363638718,
when:
No. of s model is 104.09090909041859
and
No. of y model is 10.454545454690646
